codegen/package.go

```diff
@@ -197,15 +198,114 @@ func (p PackageHelper) TypeFullName(typeSpec compile.TypeSpec) (string, error) {
 	if typeSpec == nil {
 		return "", nil
 	}
-	tfile := typeSpec.ThriftFile()
+	return goType(p, typeSpec)
+}
+
+func goType(p PackageHelper, spec compile.TypeSpec) (string, error) {
+	switch s := spec.(type) {
+	case *compile.BoolSpec:
+		return "bool", nil
+	case *compile.I8Spec:
+		return "int8", nil
+	case *compile.I16Spec:
+		return "int16", nil
+	case *compile.I32Spec:
+		return "int32", nil
+	case *compile.I64Spec:
+		return "int64", nil
+	case *compile.DoubleSpec:
+		return "float64", nil
+	case *compile.StringSpec:
+		return "string", nil
+	case *compile.BinarySpec:
+		return "[]byte", nil
+	case *compile.MapSpec:
+		k, err := goReferenceType(p, s.KeySpec)
+		if err != nil {
+			return "", err
+		}
+		v, err := goReferenceType(p, s.ValueSpec)
+		if err != nil {
+			return "", err
+		}
+		if !isHashable(s.KeySpec) {
+			return fmt.Sprintf("[]struct{Key %s; Value %s}", k, v), nil
+		}
+		return fmt.Sprintf("map[%s]%s", k, v), nil
+	case *compile.ListSpec:
+		v, err := goReferenceType(p, s.ValueSpec)
+		if err != nil {
+			return "", err
+		}
+		return "[]" + v, nil
+	case *compile.SetSpec:
+		v, err := goReferenceType(p, s.ValueSpec)
+		if err != nil {
+			return "", err
+		}
+		if !isHashable(s.ValueSpec) {
+			return fmt.Sprintf("[]%s", v), nil
+		}
+		return fmt.Sprintf("map[%s]struct{}", v), nil
+	case *compile.EnumSpec, *compile.StructSpec, *compile.TypedefSpec:
+		return goCustomType(p, spec)
+	default:
+		panic(fmt.Sprintf("Unknown type (%T) %v", spec, spec))
+	}
+}
+
+func goReferenceType(p PackageHelper, spec compile.TypeSpec) (string, error) {
+	t, err := goType(p, spec)
+	if err != nil {
+		return "", err
+	}
+
+	if isStructType(spec) {
+		t = "*" + t
+	}
+
+	return t, nil
+}
 
-	if tfile == "" {
-		return typeSpec.ThriftName(), nil
+func goCustomType(p PackageHelper, spec compile.TypeSpec) (string, error) {
+	f := spec.ThriftFile()
+	if f == "" {
+		return "", fmt.Errorf("goCustomType called with native type (%T) %v", spec, spec)
 	}
 
-	pkg, err := p.TypePackageName(tfile)
+	pkg, err := p.TypePackageName(f)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to get the full type")
+		return "", errors.Wrapf(err, "failed to get package for custom type (%T) %v", spec, spec)
 	}
-	return pkg + "." + typeSpec.ThriftName(), nil
+
+	return pkg + "." + pascalCase(spec.ThriftName()), nil
+}
+
+func isStructType(spec compile.TypeSpec) bool {
+	spec = compile.RootTypeSpec(spec)
+	_, isStruct := spec.(*compile.StructSpec)
+	return isStruct
+}
+
+// isHashable returns true if the given type is considered hashable by thriftrw.
+//
+// Only primitive types, enums, and typedefs of other hashable types are considered hashable.
+func isHashable(t compile.TypeSpec) bool {
```

Comment:
The func body says isHashable is the same as is PrimitiveType. Why keep two functions instead of only `isPrimitiveType`?